### PR TITLE
Replace `EncodePacket` and `DecodePacket` with `Packet`

### PR DESCRIPTION
**Pull request by rj00a**
## Description

Combines the `EncodePacket` and `DecodePacket` trait into a single `Packet` trait. This makes `valence_protocol` simpler and easier to use. This can be done because all packets were made to be bidirectional in #253.

Additionally, a `packet_id` method has been added. This should help with #238.

## Test Plan

Steps:
1. Run examples, packet_inspector, etc. Behavior should be the same.

### CODE REVIEW

**Review by dyc3 — APPROVED** [review]

File: crates/valence_protocol_macros/src/lib.rs [hunk 1]
```diff
@@ -8,13 +8,14 @@ use proc_macro::TokenStream as StdTokenStream;
 use proc_macro2::TokenStream;
```
Comment (JadedBlueEyes) [inline crates/valence_protocol_macros/src/lib.rs:8]: Just above here - 

```rust
//! This crate provides derive macros for [`Encode`], [`Decode`],
//! [`EncodePacket`], and [`DecodePacket`].
//! It also provides the procedural macro [`ident_str!`]
```

Broken links

File: crates/valence_protocol/src/lib.rs [hunk 2]
```diff
@@ -277,111 +278,75 @@ pub trait Decode<'a>: Sized {
 ///
 /// [macro]: valence_protocol_macros::DecodePacket
```
Comment (JadedBlueEyes) [inline crates/valence_protocol/src/lib.rs:279]: Broken link

File: crates/valence/src/unit_test/util.rs [hunk 3]
```diff
@@ -135,7 +135,7 @@ impl MockClientHelper {
 
     /// Inject a packet to be treated as a packet inbound to the server. Panics
     /// if the packet cannot be sent.
-    pub fn send(&mut self, packet: &(impl EncodePacket + ?Sized)) {
+    pub fn send<'a>(&mut self, packet: &(impl Packet<'a>)) {
```
Comment (JadedBlueEyes) [inline crates/valence/src/unit_test/util.rs:138]: unnecessary parentheses around type

File: crates/valence_protocol/src/lib.rs [hunk 4]
```diff
@@ -396,7 +361,7 @@ mod derive_tests {
     #[allow(unconditional_recursion)]
     fn has_impls<'a, T>()
```
Comment (JadedBlueEyes) [inline crates/valence_protocol/src/lib.rs:362]: type parameter goes unused in function definition

**Review by JadedBlueEyes — CHANGES_REQUESTED** [review]
Hey, I've just gone through and ran clippy + tests. I've marked out a few changes, and the playground is broken from a previous PR, but otherwise LGTM! :+1: :smile:

File: crates/valence_protocol/src/lib.rs (hunk 4 above)
Comment (rj00a) [reply crates/valence_protocol/src/lib.rs:362]: The type param asserts that the input types have the expected bounds.

**Pull request merged**